### PR TITLE
Fixed #1586

### LINKED ISSUE #1586: 关于正则路由 带括号问题
正则路由里面配置// @router /:app((a|b|c)) [get]

发现编译多几次后~

golang会报错

panic: regexp: Compile(`^(([a|b|c])(([a|b|c]))$`): error parsing regexp: missing closing ): `^((a|b|c)((a|b|c))$`

如果正则需要带括号该怎么办呢？

### PR DESCRIPTION
**Pull request by ysqi**
URL support special characters `()` ,e.g:
```go
beego.Router(`/v1/:appid\((a|b|c)\)`, &controllers.RController{})
```
 match:  `/v1/001(a)` , :appid=001

### CODE REVIEW

File: tree.go [hunk 1]
```diff
@@ -15,12 +15,11 @@
 package beego
 
 import (
+	"github.com/astaxie/beego/context"
+	"github.com/astaxie/beego/utils"
```
Comment (astaxie) [inline tree.go]: why move these two line to the top. it should be follow the goimport

Comment (ysqi) [inline tree.go]: 知道原因了，原来是被我删除了中间的空行，导致go fmt 移动位置 了，我恢复下

File: tree_test.go [hunk 2]
```diff
@@ -245,48 +249,31 @@ func TestSplitPath(t *testing.T) {
...
+		":id":                        {true, []string{":id"}, ""},
+		"?:id":                       {true, []string{":", ":id"}, ""},
+		":id:int":                    {true, []string{":id"}, "([0-9]+)"},
+		":name:string":               {true, []string{":name"}, `([\w]+)`},
+		":id([0-9]+)":                {true, []string{":id"}, `([0-9]+)`},
+		":id([0-9]+)_:name":          {true, []string{":id", ":name"}, `([0-9]+)_(.+)`},
+		":id(.+)_cms.html":           {true, []string{":id"}, `(.+)_cms.html`},
+		"cms_:id(.+)_:page(.+).html": {true, []string{":id", ":page"}, `cms_(.+)_(.+).html`},
+		`:app(a|b|c)`:                {true, []string{":app"}, `(a|b|c)`},
+		`:app\((a|b|c)\)`:            {true, []string{":app"}, `(.+)\((a|b|c)\)`},
```
Comment (astaxie) [inline tree_test.go]: I like this improvement. Thanks.

Comment (ysqi) [inline tree_test.go]: I'm very happy that you liked it !

**Comment by cjsky** [comment]
very good !!

**Pull request merged**